txpool/txpool.go

```diff
@@ -192,576 +200,280 @@ func NewTxPool(
-	evnt := &blockchain.Event{
-		NewChain: headers,
-	}
+// Demote pops the recoverable transaction from
+// its associated promoted queue (account) and
+// issues an enqueueRequest for it.
+// Will update executables with the next primary
+// from that account (if any).
+func (p *TxPool) Demote(tx *types.Transaction) {
+	account := p.accounts.from(tx.From)
 
-	// Process the txns in the event to make sure the TxPool is up-to-date
-	t.ProcessEvent(evnt)
-}
+	account.promoted.lock(true)
+	defer account.promoted.unlock()
 
-// processEventWrapper holds metadata information related
-// to a new block insertion event that relate to a specific account
-type processEventWrapper struct {
-	stateNonce   uint64
-	transactions []*types.Transaction
-}
+	// drop the tx from account promoted
+	account.promoted.pop()
 
-// addTxn adds a transaction to the account's process event wrapper
-func (p *processEventWrapper) addTxn(txn *types.Transaction) {
-	p.transactions = append(p.transactions, txn)
-}
+	// signal enqueue request [BLOCKING]
+	p.enqueueReqCh <- enqueueRequest{tx: tx, demoted: true}
 
-// promotedTxnCleanup looks through the promoted queue for any invalid transactions
-// made by a specific account, and removes them
-func (t *TxPool) promotedTxnCleanup(
-	address types.Address, // The address to filter by
-	stateNonce uint64, // The valid nonce (reference for pruning)
-	cleanupCallback func(txn *types.Transaction), // Additional cleanup logic
-) {
-	// Prune out all the now possibly low-nonce transactions in the promoted queue
-	t.pendingQueue.lock.Lock()
-
-	// Find the txns that correspond to this account
-	droppedPendingTxs := make([]*types.Transaction, 0)
-	for _, pendingQueueTxn := range t.pendingQueue.index {
-		// Check if the txn in the promoted queue matches the search criteria
-		if pendingQueueTxn.from == address && // The sender of this txn is the account we're looking for
-			pendingQueueTxn.tx.Nonce < stateNonce { // The nonce on this promoted txn is invalid
-			// Transaction found, drop it from the pending queue
-			if dropped := t.pendingQueue.dropTx(pendingQueueTxn.tx); dropped {
-				// Save the transaction as dropped
-				droppedPendingTxs = append(droppedPendingTxs, pendingQueueTxn.tx)
-			}
-		}
+	// update executables
+	if tx := account.promoted.peek(); tx != nil {
+		p.executables.push(tx)
 	}
 
-	t.pendingQueue.lock.Unlock()
-
-	for _, droppedPendingTx := range droppedPendingTxs {
-		// Update the log
-		t.logger.Debug(
-			fmt.Sprintf(
-				"Dropping promoted txn [%s]",
-				droppedPendingTx.Hash.String(),
-			),
-		)
+	p.logger.Debug("demoted transaction", "hash", tx.Hash.String())
+}
 
-		cleanupCallback(droppedPendingTx)
+// ResetWithHeaders is called from within ibft when the node
+// has received a new block from a peer. The pool needs to align
+// its own state with the new one so it can correctly process
+// further incoming transactions.
+func (p *TxPool) ResetWithHeaders(headers ...*types.Header) {
```

Comment:
I felt like writing a unit test for this method would be overkill, since all the important logic for this call is happening in `resetAccounts`. Everything before that is just parsing the `headers` into a map of nonces (`map[type.Address]uint64`).